### ISSUE TITLE
Chat with System (NLP)

### ISSUE THREAD

**Issue by dodoels** [Cyberjusticelab/JusticeAI#8]
As a user, I want to express an issue I'm facing using natural language

*Scope of work*
- [ ] Determine whether user input is relevant (12 hours)
- [ ] Clarify micro-service API endpoints (2 hours)
- [ ] In ambiguous situations, confirm input with user (1 hours)
- [ ] Determine the answer to a yes or no response to a question by the user (6 hours)
- [ ] Determine the classification of an answer into a given category (6 hours)
- [ ] Determine what the primary fact is within a response to a question by the user (15 hours)

*Acceptance Criteria*
- System asks a yes/no question. User gives a response describing a yes/no sentiment. System is able to classify the response among specific response types.
- System asks a question requiring a long answer. User gives a textual response. System is able to parse the response and classify the response among specific response types.

*Story Points*
- 20

*Priority*
- High

*Risk*
- High

**Comment by lancelafontaine** [comment]
*Scope of work*
- [ ] Prompt user to enter name (? hours)
- [ ] Determine whether user input is relevant (? hours)
- [ ] In ambiguous situations, confirm input with user (? hours)
- [ ] Determine inconsistencies between responses (? hours)

**Comment by lancelafontaine** [comment]
![20170929_193506](https://user-images.githubusercontent.com/1548285/31039682-a1bd3d32-a54d-11e7-8c67-7feeb89c00d9.jpg)

**Comment by Samuel-Campbell** [comment]
-----------------------------------------
Standford Library
-----------------------------------------
PROS:
- Supports French and English
- Better runtime than NLTK

CONS:
- GPL license conflicts with our MIT license

-----------------------------------------
SpacY
-----------------------------------------
PROS:
- Better performance overall than NLTK. (Word tagging)

CONS:
- Only supports english

-----------------------------------------
Textblob
-----------------------------------------
Built over NLTK and pattern libraries

PROS:
- Better runtime than NLTK
- Good for very simple NLP operations
- Implements NLTK and pattern efficiently together

CONS:
- Functionality is more limited and less options for customization

-----------------------------------------
Gensim
-----------------------------------------
PROS: 
- Good for finding document similarities (not very useful to us)
- word2vec

CONS: 
- only proficient at finding document similarities and offers limited other functionaliteis
- GNU license

-----------------------------------------
Pattern
-----------------------------------------
PROS:
- Better runtime than NLTK
- Efficient at speech tagging and word tagging

CONS: 
- only supported for python 2

-----------------------------------------
NLTK
-----------------------------------------
PROS:
- General purpose NLP good for learning, supports standford library addon.
- Apache lisence
- Supports many functionalities and is compatible with tensorflow, scitoolkit and standford extensions
- Alot of online support
- Can download french tokenizer online

CONS:
- Worst runtime of any NLP library

-----------------------------------------
TreeTagger
-----------------------------------------
PROS:
- Can tag words in french and english

CONS:
- GNU lisence

-----------------------------------------
GLOVE
-----------------------------------------
In Glove it tries to capture the counts of overall statistics how often its appears.
co-occurrence matrix for the entire corpus first, then factorize it to yield matrices for word vectors and context vectors.

PROS:
- Apache lisence
- Better alternative than Word2Vec from gensim

CONS:

-----------------------------------------
Word2vec
-----------------------------------------
"Stream through n-grams of words, attempting to train a neural network to predict the n-th word given words [1,...,n-1] or the other way round. The end result is a matrix of word vectors or context vectors respectively." - Stackoverflow guy

**Comment by Samuel-Campbell** [comment]
---------------------------------------------------
REQUIRED LIBRARIES
---------------------------------------------------

- Numpy
- Matplotlib
- NLTK
- Glove
- scitools / tensorflow (arek edify me on this)
- pymysql

**Comment by arekmano** [comment]
The following are the tentative API endpoints:

- /introduction
  - input:
> {
>   "person" : the text response to the answer of whether or not they are a tenant or a landlord,
>   "conversationID" : the unique ID of the conversation, unused for now
>   "name" : the text response containing the user's name
> }
  - output:
> {
>   "personClass" : the user's class. Either 'tenant' or 'landlord',
>   "conversationID" : the unique ID of the conversation, same as input
>   "name" : the user's name string.
> }

- /yesno
  - input
> {
>   "conversation_id" : the unique ID of the conversation, unused for now,
>   "answer" : the text response containing a yes/no answer,
>   "question_id" : the question to which the user is replying. Unused
> }
  - output:
>  {
>   "question_id" : the question to which the user is replying. Unused
>    "facts" : an array containing all of the facts that we established from answer. First element is always either 'positive' or 'negative' for yes/no answer
>   "conversation_id" : the unique ID of the conversation, unused for now,
>  }

**Comment by Samuel-Campbell** [comment]
--------------------------------------------------------------
- **Fact Extraction:**
// noun phrase
Clause: {
    Word
    adjective
    quantifier --> [A] cat has four legs.  (Translation) --> There exists a cat with four legs
}
// verb phrase, single preposition
Predicate: {
    word
    adverb --> He [really] did well
}
// Adjective phrase, adverb phrase, noun phrase
Compliment/Clause:{
    word
    adjective
    quantifier
}

--------------------------------------------------------------
- **Sentence types:**

Adverb phrase: [With great regret], I announce Mihai's lost.
Adjective phrase: [Grass-fed organic] beef is the best choice.
Noun phrase: [The cat] is blue.
verb phrase: He [has been dragging] my balls.
Preposition phrase: The man walked the dog [in the forest]

---------------------------------------------------------------------
- **Analysing a sentence**

The man with a dog was in the forest.

[Noun phrase: The man ]
[preposition: with]
[Noun phrase: a dog]

[Noun phrase: the man]
[verb phrase: was]
[preposition phrase: in [Noun phrase: the forest]]

**Comment by arekmano** [comment]
Updated API


- /problem_category
  - input
> {
>   "conversation_id" : the unique ID of the conversation, UNUSED
>   "answer" : the text response containing a yes/no answer,
>   "question_id" : the question to which the user is replying. UNUSED
> }
  - output:
>  {
>   "question_id" : the question to which the user is replying. UNUSED
>    "facts" : [{"category" : (one of the following 'deposits', 'lease_termination', 'nonpayment', 'rent_change')}]
>   "conversation_id" : the unique ID of the conversation, UNUSED
>  }

- /fact_extraction
  - input
>  {
>    "answer" : "Answer String given by the user",
>    "facts" : An array of the facts to look for e.g. ['lease_term_type','has_lease_expired','is_tenant_dead','is_student','is_habitable','lease_term_length','lease_ends_in']
>  }
  - output
>  {
>   "question_id" : the question to which the user is replying. UNUSED
>    "facts" : array of key/value pairs of facts detected e.g. [{'lease_term_type': 'fixed' or 'indeterminate' },{'has_lease_expired': true },{'is_tenant_dead': true },{'is_student': true },{'is_habitable': true },{'lease_term_length': 1 }, {'lease_ends_in': 2 }]
>   "conversation_id" : the unique ID of the conversation, UNUSED
>  }

**Comment by naregeff** [comment]
Functionality is demonstrated. Requires more datasets for desired accuracy. Strive for 85% classification accuracy. Provide opportunity for continuous improvement as users correct misclassified responses.

**Comment by arekmano** [comment]
Create new user story with increased classification precision + recall.

**Comment by arekmano** [comment]
Created a new user story for tenant/landlord problem classification at #75
